ft-email-auth-provider/src/handlers/user_data_by_code.rs

```diff
@@ -0,0 +1,56 @@
+#[derive(serde::Serialize)]
+struct PreFillData {
+    email: String,
+}
+
+/// Get user data from `code`. This code is sent to user's email for promotions.
+#[ft_sdk::data]
+fn user_data_by_code(
+    mut conn: ft_sdk::Connection,
+    ft_sdk::Query(code): ft_sdk::Query<"code">,
+) -> ft_sdk::data::Result {
+    use diesel::prelude::*;
+
+    #[derive(diesel::QueryableByName)]
+    #[diesel(table_name = ft_sdk::auth::fastn_user)]
+    struct UD {
+        data: String,
+    }
+
+    let data = match diesel::sql_query(
+        r#"
+            SELECT
+                data -> 'email' as data
+            FROM fastn_user
+            WHERE
+                EXISTS (
+                    SELECT 1
+                    FROM json_each ( data -> 'subscription' -> 'subscription' -> 'confirmation-code')
+                    WHERE value = $1
+                )
+        "#,
+    )
+    .bind::<diesel::sql_types::Text, _>(&code)
+    .get_result::<UD>(&mut conn)
+    {
+        Ok(ud) => ud.data,
+        Err(diesel::result::Error::NotFound) => {
+            return Err(ft_sdk::single_error("email", "").into());
```

Comment:
Write proper error message.

Reply:
done in https://github.com/fastn-community/auth/pull/7/commits/7e2a68e84dbcc778dc167caab1eb4cbef7dbdb2b